comp/core/agenttelemetry/impl/sender.go

```diff
@@ -268,61 +269,73 @@ func (s *senderImpl) addMetricPayload(
 
 func (s *senderImpl) startSession(cancelCtx context.Context) *senderSession {
 	return &senderSession{
-		cancelCtx: cancelCtx,
+		cancelCtx:       cancelCtx,
+		payloadTemplate: s.payloadTemplate,
 	}
 }
 
-func (s *senderImpl) flushSession(ss *senderSession) error {
-	// There is nothing to do if there are no payloads
-	if len(ss.payloads) == 0 {
-		return nil
-	}
-
-	s.logComp.Infof("Flushing Agent Telemetery session with %d payloads", len(ss.payloads))
-
-	// Defer cleanup of payloads. Even if there is an error, we want to cleanup
-	// but in future we may want to add retry logic.
+func (ss *senderSession) flush() Payload {
 	defer func() {
-		ss.payloads = nil
+		// Clear the payloads
+		ss.metricPayloads = nil
 	}()
 
 	// Create a payload with a single message or batch of messages
-	payload := s.payloadTemplate
+	payload := ss.payloadTemplate
 	payload.EventTime = time.Now().Unix()
-	if len(ss.payloads) == 1 {
+	if len(ss.metricPayloads) == 1 {
 		// Single payload will be sent directly using the request type of the payload
-		payload.RequestType = ss.payloads[0].requestType
-		payload.Payload = ss.payloads[0].payload
-	} else {
-		// Batch up multiple payloads into single "batch" payload type
-		payload.RequestType = "message-batch"
-		payloadWrappers := make([]BatchPayloadWrapper, 0)
-		for _, p := range ss.payloads {
-			payloadWrappers = append(payloadWrappers,
-				BatchPayloadWrapper{
-					RequestType: p.requestType,
-					Payload:     p.payload,
-				})
-		}
-		payload.Payload = payloadWrappers
+		mp := ss.metricPayloads[0]
+		payload.RequestType = "agent-metrics"
+		payload.Payload = payloadInfo{"agent-metrics", mp}.payload
```

Comment:
Out of curiosity, what is the reason to have a special case for 1 metric. Can't it be able using `"message-batch"`?

Reply:
Why, it sends a bit more data and on the backend they need to unwrap it to.  And if you have millions of Agents it will add up